sticker/src/tensorflow/lr.rs

```diff
@@ -40,8 +45,11 @@ impl LearningRateSchedule for ConstantLearningRate {
 /// *lr = initial_lr * decay_rate ^ (epoch / decay_steps)*
 pub struct ExponentialDecay {
     initial_lr: f32,
+    lr: f32,
     decay_rate: f32,
     decay_steps: usize,
+    warmup_steps: usize,
+    step: usize,
```

Comment:
`step` and `warmup_steps` get confusing here, because `decay_steps` refers to the number of epochs, while `step` and `warmup_steps` refer to batches?

Reply:
I think decay steps should actually refer to batches too (e.g. https://www.tensorflow.org/api_docs/python/tf/compat/v1/train/exponential_decay). In TF terminology a step is usually a single call to the train op.

Reply:
A lot of this code has a heritage from before Tensorflow (through the old Go versions of dparnn/toponn). I agree that using batches would be better, since it would decay in a more fine-grained manner. However, it would also mean that the meaning of the parameters related to this schedule would change.

At any rate, I think words like *iteration* or *step* are very confusing, especially since we have the non-ambiguous terms *batch* and *epoch*. So I propose to rename `warmup_steps` and `step` to `warmup_batches` and `batch`. We can then later change the behavior of decay to be in number of batches as well.

Reply:
When I read batch, I'd expect it to be about the n-th batch in the epoch and not about the total number of processed batches.

In tensorflow and pytorch step describes a single iteration in the optimization process:

* the `global_step` argument of `tf.train.AdamOptimizer.minimize` (https://www.tensorflow.org/api_docs/python/tf/compat/v1/train/AdamOptimizer#minimize) or `tf.train.exponential_decay` (https://www.tensorflow.org/api_docs/python/tf/compat/v1/train/exponential_decay)
* in pytorch applying the gradients happens via calling `step` on the optimizer (https://pytorch.org/docs/stable/optim.html)

Would `global_step` or `global_train_step` be an acceptable name?

Reply:
Not happy about `decay_epochs` either. If `staircase == true` it describes the width of a "step" in the decay stairs but more importantly it also influences the slope. Both properties don't really seem to be reflected in `decay_steps` or `decay_epochs` which I would either read as _decay-steps-before-zero-lr_ or _decay-steps-before-constant-lr_.

Unfortunately I cannot come up with a better name either. For now, I'll use `decay_epochs` to make clear it's about epochs, not steps. Maybe we'll come up with something better when changing the decay-behavior.